{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import random\n",
    "import DNA\n",
    "import mutation\n",
    "from DNA import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on module mutation:\n",
      "\n",
      "NAME\n",
      "    mutation - This module is for various mutations of sequence data.\n",
      "\n",
      "DESCRIPTION\n",
      "    Functions include random RNA generator, mutation programs that modify the RNA strings\n",
      "    and mutations finder programs that attempt to identify the mutation.\n",
      "    \n",
      "    Copyright Benjamin Clyde, 2017\n",
      "\n",
      "FUNCTIONS\n",
      "    deletion(rna)\n",
      "        Function to take an RNA string and randomly choose an index for the string and then\n",
      "        randomly delete 1-9 nucleotides sequentially and return the new RNA string\n",
      "    \n",
      "    find_codon_diff(protein, mutated_protein)\n",
      "        Function to find the effect the SNP had on the protein.\n",
      "        First compares the length of the proteins to see if the SNP caused a premature\n",
      "        stop codon, otherwise, returns the new codon.\n",
      "    \n",
      "    find_deletion(rna, deleted_rna)\n",
      "        Function to find the deleted sequence from an RNA seq\n",
      "        First takes the difference in length between the two sequences to find the amount\n",
      "        of nucleotides that were inserted, then compare the sequences until it can determine\n",
      "        where the sequences differed.\n",
      "    \n",
      "    find_insertion(rna, inserted_rna)\n",
      "        Function to find the inserted sequence into an RNA seq\n",
      "        First takes the difference in length between the two sequences to find the amount\n",
      "        of nucleotides that were inserted, then compare the sequences until it can determine\n",
      "        where the sequences differed.\n",
      "    \n",
      "    find_mutation(rna, mutated_rna)\n",
      "        Function to take an RNA string and the function mutated string to find the mutation\n",
      "        regardless of type. Limited to finding one mutation at a time, will look into\n",
      "        multiple mutations in a later version\n",
      "    \n",
      "    find_snp(rna, mutated_rna)\n",
      "        Function compares the old sequence to the mutated sequence to find the SNP\n",
      "        Returns the position, the previous nucleotide and the new nucleotide\n",
      "    \n",
      "    insertion(rna)\n",
      "        Function to take an RNA string and randomly choose an index for the string and then\n",
      "        randomly add 1-9 nucleotides sequentially and return the new RNA string\n",
      "    \n",
      "    mutator(rna)\n",
      "        Function to take an RNA string and randomly apply one of the mutation functions.\n",
      "        Probably should adjust the random to closely match the frequency the mutations\n",
      "        occur in nature, but couldn't find good numbers.\n",
      "    \n",
      "    random_rna(length)\n",
      "        Function to take a given length and return a random RNA sequence\n",
      "        Function ensures the RNA sequence starts with a start codon and ends with a stop codon\n",
      "    \n",
      "    snp(sequence)\n",
      "        Function takes a sequence string and introduces a single nucleotide polymorphism (SNP)\n",
      "        Randomly finds an index in the string and replaces the value at the index with a new\n",
      "        letter from the given alphabet. If the value is the same as the previous value, the\n",
      "        function calls itself again to make sure the sequence changes.\n",
      "\n",
      "DATA\n",
      "    rna_codon_list_no_stop = ['GCG', 'UCG', 'GCC', 'CAA', 'GAU', 'AGA', 'G...\n",
      "    rna_codon_table = {'AAA': 'K', 'AAC': 'N', 'AAG': 'K', 'AAU': 'N', 'AC...\n",
      "\n",
      "FILE\n",
      "    /home/jovyan/work/beniclyde_6018_2017/DNA Manipulation/mutation.py\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(mutation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "rna1 = mutation.random_rna(99)\n",
    "rna2 = mutation.random_rna(99)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'AUGUACACUAUCACUCUCGCCCCCACCACGUGGGGCAGCUUACAAACCAAUGGGUCAGAGUGGCACACCUCAGGCGAGCACGUCUCGAUGAACAGGUGA'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rna1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'AUGUGGAUCCGCUUUGGACUAGUCGAUAGGAACCCUAGUGCCUUCCAUCUGAGAUUGAUAUGCGGGGCGAUAACCCAACGUAGCUUGGCUAGUCUCUGA'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rna2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def dotplot_list_of_lists(dna_x, dna_y):\n",
    "    dotplot_matrix = [['0' for x in dna_x] for y in dna_y]\n",
    "    for x_index, x_value in enumerate(dna_x):\n",
    "        for y_index, y_value in enumerate(dna_y):\n",
    "            if x_value == y_value:\n",
    "                dotplot_matrix[y_index][x_index] = '1'\n",
    "    return dotplot_matrix\n",
    "\n",
    "dna_x = 'TAATGCCTGAAT'\n",
    "dna_y = 'CTCTATGCC'\n",
    "\n",
    "M = dotplot_list_of_lists(dna_x, dna_y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[1 0 0 ..., 0 1 1]\n",
      " [0 1 1 ..., 1 0 0]\n",
      " [0 1 1 ..., 1 0 0]\n",
      " ..., \n",
      " [0 1 1 ..., 1 0 0]\n",
      " [1 0 0 ..., 0 1 1]\n",
      " [1 0 0 ..., 0 1 1]]\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXQAAAD8CAYAAABn919SAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAErhJREFUeJzt3W+InWeZx/Hvz7aiqV3bbLMhWNu4UCpFbCuzXUUR7R+p\nXdmWfSGKSlgKeeMuFQRNERZ8UbavRF8sQvBfYLNKqboNxVViVBZB1IlW7T83rtvUlqQZkxW3FtTa\na1+cZ3QyPZNzZs7f5znfDwznPPd5pnPfpLnmznXu61ypKiRJ7feiWU9AkjQeBnRJ6ggDuiR1hAFd\nkjrCgC5JHWFAl6SOMKBLUkcY0CWpIwzoktQR50/zh1166aW1e/fuaf5ISWq9o0eP/rKqdgy6b6oB\nfffu3SwvL0/zR0pS6yU5Psx9plwkqSOGCuhJLk5yX5LHkjya5A1Jtic5nORY83jJpCcrSdrYsDv0\nTwBfrapXA9cAjwL7gCNVdSVwpLmWJM3IwICe5OXAm4FPA1TV76rqV8BtwIHmtgPA7ZOapCRpsGF2\n6K8CVoDPJvlhkk8luRDYWVUnmntOAjsnNUlJ0mDDBPTzgdcBn6yq64DfsC69Ur0uGX07ZSTZm2Q5\nyfLKysqo85Wkdjl4EHbvhhe9qPd48ODEftQwAf1J4Mmq+m5zfR+9AP90kl0AzeOpft9cVfuraqmq\nlnbsGHiMUpK64+BB2LsXjh+Hqt7j3r0TC+oDA3pVnQR+keSqZuhG4BHgELCnGdsD3D+RGUpSW33k\nI/Dss2ePPftsb3wChi0s+kfgYJIXAz8H/p7eL4N7k9wBHAfeOZEZSlJbPfHE5sZHNNSxxap6sEmb\nvLaqbq+q/62q01V1Y1VdWVU3VdWZicxQktqgX6788sv737vR+IisFJWkUW2UK7/1Vti27ex7t22D\nu++eyDQM6JI0qo1y5V/5CuzfD1dcAUnvcf9+eM97JjINA7okbdb69MrxDT4764knesH78cfh+ed7\njxMK5jDlT1uUpNZbTa+s7siPH+/tvqtPKc6EcuUbcYcuSZvRL71S1Qvqa00wV74RA7okncuw6ZWq\nqeXKN2LKRZI2spn0yhVX9HLkM+QOXZI2MsfplX4M6JK0kY0qOucgvdKPAV2SVq3Pl2/f3v++1fTK\nFI4iboY5dEmC/vnyCy6AF78Yfve7P903J+mVftyhSxL0z5f//vdw0UVzmV7pxx26pMV08GAviD/x\nRK8AaKPjiGfOwC9/Od25bZEBXdLimeNqz1GYcpG0eFp2HHFYBnRJ3deias9RmHKR1G0tq/YchTt0\nSd3W0fRKPwZ0Sd3WsmrPURjQJXXHZvp6zmm15ygM6JK6YU76es6SAV1SN8xJX89ZMqBLaqc57es5\nSx5blNQ+Ha30HJU7dEnts0BHETdjqICe5PEkP0nyYJLlZmx7ksNJjjWPl0x2qpIW1oJUeo5qMymX\nt1bV2o8c2wccqap7kuxrrj881tlJ0gJVeo5qlJTLbcCB5vkB4PbRpyNJ65heGdqwAb2Aryc5mmRv\nM7azqk40z08CO/t9Y5K9SZaTLK+srIw4XUkLZ4EqPUc1bMrlTVX1VJK/AA4neWzti1VVSfr8+weq\naj+wH2BpaanvPZL0R+sbT2zfDqdPv/C+BU+v9DNUQK+qp5rHU0m+DFwPPJ1kV1WdSLILODXBeUpa\nBB3o6zlLA1MuSS5MctHqc+BtwEPAIWBPc9se4P5JTVLSguhAX89ZGmaHvhP4cnpvQJwP/FtVfTXJ\n94F7k9wBHAfeOblpSuqkDvb1nKWBAb2qfg5c02f8NHDjJCYlaQFY7Tl2VopKmg2PI46dAV3SdFjt\nOXF+OJekybPacyrcoUuaPNMrU2FAlzR5VntOhQFd0ngteF/PWTKgSxof+3rOlAFd0vjY13OmDOiS\nts6+nnPFY4uStsZKz7njDl3S1ngUce4Y0CUNx0rPuWfKRdJgVnq2gjt0SYOZXmkFA7qkwaz0bAUD\nuqQXWp8v3769/31Wes4Vc+iSzmZfz9Zyhy7pbPb1bC136NKis69nZxjQpUVmtWenmHKRFpnHETvF\ngC4tMo8jdooBXVokHkfsNHPo0qLwOGLnDb1DT3Jekh8meaC53p7kcJJjzeMlk5umpJF5HLHzNpNy\nuRN4dM31PuBIVV0JHGmuJc2LYT8d8cwZ0ysdMVRAT3IZ8DfAp9YM3wYcaJ4fAG4f79QkbVm/3p7r\nT66s8jhiZwy7Q/848CHg+TVjO6vqRPP8JLBznBOTNAKPIy6kgQE9yTuAU1V1dKN7qqqAPpUIkGRv\nkuUkyysrK1ufqaT+1qdWDh70OOKCSvWrCFt7Q/LPwPuA54CXAH8GfAn4K+AtVXUiyS7gW1V11bn+\nW0tLS7W8vDyWiUvihSdXoLfrfulL4fTpF95v84lWSnK0qpYG3Tdwh15Vd1XVZVW1G3gX8I2qei9w\nCNjT3LYHuH+E+Urain6pldXrbdvOHje90nmjFBbdA9yc5BhwU3MtaZo2Sq2cOdNLp5heWSgDUy7j\nZMpFGtH6T0Z85hlTKwtg2JSLlaJSW1jpqQH8LBepLaz01ADu0KV5ZeMJbZIBXZpHNp7QFphykeaR\nlZ7aAgO6NA+G/SAtKz11DqZcpFnbTHrF44g6B3fo0qyZXtGYGNClWfODtDQmBnRp2uzrqQkxhy5N\nk9WemiB36NI0We2pCXKHLk2S1Z6aIgO6NClWe2rKTLlIk+JxRE2ZAV0aB/t6ag6YcpFG1S+1sndv\n7ziizSc0Re7QpVHZ11NzwoAujcq+npoTBnRps4at9Lz88l7wttpTU2IOXdoMKz01x9yhS5thpafm\nmDt06Vys9FSLGNCljVjpqZYx5SJtxEpPtczAgJ7kJUm+l+RHSR5O8tFmfHuSw0mONY+XTH660gTZ\n11MtN0zK5bfADVX1TJILgG8n+Q/g74AjVXVPkn3APuDDE5yrNDn29VQHDNyhV88zzeUFzVcBtwEH\nmvEDwO0TmaE0DaZX1AFD5dCTnJfkQeAUcLiqvgvsrKoTzS0ngZ0bfO/eJMtJlldWVsYyaWns/CAt\ndcBQAb2q/lBV1wKXAdcnec2614verr3f9+6vqqWqWtqxY8fIE5bGwr6e6qBNHVusql8l+SZwC/B0\nkl1VdSLJLnq7d2n+We2pjhrmlMuOJBc3z18K3Aw8BhwC9jS37QHun9QkpbGy2lMdNcwOfRdwIMl5\n9H4B3FtVDyT5DnBvkjuA48A7JzhPaeus9tSCGBjQq+rHwHV9xk8DN05iUtLYWO2pBWKlqLrN44ha\nIAZ0dYd9PbXg/HAudYN9PSV36OoI+3pKBnR1hH09JQO6Wsq+ntILmENX+1jpKfXlDl3tY6Wn1Jc7\ndM0/Kz2loRjQNd+s9JSGZspF881KT2loBnTNNys9paEZ0DVfbDwhbZk5dM0PjyNKI3GHrvnhcURp\nJO7QNTseR5TGyoCu2fA4ojR2plw0Gx5HlMbOgK7pWH96ZaP0iscRpS0z5aLJ20x6xcYT0pa5Q9fk\nmV6RpsKArsmz2lOaCgO6xs9qT2kmzKFrvKz2lGZm4A49ySuTfDPJI0keTnJnM749yeEkx5rHSyY/\nXc09qz2lmRkm5fIc8MGquhp4PfD+JFcD+4AjVXUlcKS51iJZn1o5ePDczZpNr0gTNTCgV9WJqvpB\n8/z/gEeBVwC3AQea2w4At09qkppDq6mV48d7b24eP967PlezZkkTtak3RZPsBq4DvgvsrKoTzUsn\ngZ1jnZnmW7/Uyur1tm1nj5svl6Zi6ICe5GXAF4EPVNWv175WVQX0qRKBJHuTLCdZXllZGWmymqFh\nKz3PnOnlx82XS1OX6lett/6m5ALgAeBrVfWxZuynwFuq6kSSXcC3quqqc/13lpaWanl5eQzT1lSt\nP7kCVnpKU5TkaFUtDbpvmFMuAT4NPLoazBuHgD3N8z3A/VuZqFrASk+pFYZJubwReB9wQ5IHm69b\ngXuAm5McA25qrtVFVnpKrTCwsKiqvg1kg5dvHO90NBfWN57Yvh1On37hfaZXpLlipajOZqWn1Fp+\nlovOZqWn1Fru0BedfT2lzjCgLzL7ekqdYsplkXkcUeoUA/oisa+n1GmmXBaFfT2lznOHvihMr0id\nZ0BfFFZ7Sp1nQO8q+3pKC8ccehdZ7SktJHfoXWS1p7SQDOhtZ19PSQ1TLm3WL7Wy2tez36cjWu0p\ndZo79Dazr6ekNQzobWJfT0nnYMqlLTb7QVrveY8BXFow7tDbwkpPSQMY0NvCSk9JAxjQ55WVnpI2\nyRz6PLLSU9IWuEOfR1Z6StoCd+jzwL6eksbAgD5r9vWUNCamXGbN44iSxmRgQE/ymSSnkjy0Zmx7\nksNJjjWPl0x2mh1iX09JEzLMDv1zwC3rxvYBR6rqSuBIc61BVtMrx4/3AvZqeqUfjyNK2qSBAb2q\n/hM4s274NuBA8/wAcPuY59VNplckTdBWc+g7q+pE8/wksHOjG5PsTbKcZHllZWWLP64jrPaUNEEj\nvylaVQX0OZLxx9f3V9VSVS3t2LFj1B/XLlZ7SpqirR5bfDrJrqo6kWQXcGqck+oEqz0lTdlWd+iH\ngD3N8z3A/eOZTodY7SlpyoY5tvh54DvAVUmeTHIHcA9wc5JjwE3N9eKyr6ekOTAw5VJV797gpRvH\nPJd2sq+npDlhpeio7OspaU4Y0DfLvp6S5pQfzrUZ9vWUNMfcoW+GlZ6S5pgBfTOs9JQ0xwzo52Kl\np6QWMYe+ESs9JbWMO/SNWOkpqWXcoa+yr6ekljOgg309JXWCKRfwOKKkTjCgg8cRJXXCYgZ0jyNK\n6qDFy6F7HFFSRy3eDt3jiJI6qvs7dI8jSloQ3Q7oHkeUtEC6nXLxOKKkBdKdgL6Zvp4eR5TUQd1I\nuWy2r+fqcURJ6pBu7NDt6ylJHQnoG6VW7OspaYG0M6APW+m52tfTak9JC6B9OXQrPSWpr5F26Elu\nSfLTJD9Lsm9ckzonKz0lqa8t79CTnAf8C3Az8CTw/SSHquqRcU0OsNJTkoY0SsrleuBnVfVzgCRf\nAG4DxhfQrfSUpKGNknJ5BfCLNddPNmPjY6WnJA1t4qdckuxNspxkeWVlZXPfbKWnJA1tlJTLU8Ar\n11xf1oydpar2A/sBlpaW+uRKzmGjnLmVnpL0AqPs0L8PXJnkVUleDLwLODSeaTXuvttKT0ka0pYD\nelU9B/wD8DXgUeDeqnp4XBMDemkUKz0laSipfidGJmRpaamWl5en9vMkqQuSHK2qpUH3tbP0X5L0\nAgZ0SeoIA7okdYQBXZI6woAuSR0x1VMuSVaADT5da6BLgS59+laX1tOltYDrmWddWgsMv54rqmrH\noJumGtBHkWR5mGM7bdGl9XRpLeB65lmX1gLjX48pF0nqCAO6JHVEmwL6/llPYMy6tJ4urQVczzzr\n0lpgzOtpTQ5dknRubdqhS5LOoRUBfSbNqMcoyWeSnEry0Jqx7UkOJznWPF4yyzkOK8krk3wzySNJ\nHk5yZzPeuvUkeUmS7yX5UbOWjzbjrVvLWknOS/LDJA80161dT5LHk/wkyYNJlpuxVq4nycVJ7kvy\nWJJHk7xh3GuZ+4C+phn124GrgXcnuXq2s9q0zwG3rBvbBxypqiuBI811GzwHfLCqrgZeD7y/+fNo\n43p+C9xQVdcA1wK3JHk97VzLWnfS+0jrVW1fz1ur6to1x/vaup5PAF+tqlcD19D7MxrvWqpqrr+A\nNwBfW3N9F3DXrOe1hXXsBh5ac/1TYFfzfBfw01nPcYvruh+4ue3rAbYBPwD+us1rodc57AhwA/BA\nM9bm9TwOXLpurHXrAV4O/A/N+5aTWsvc79CZRjPq2dhZVSea5yeBnbOczFYk2Q1cB3yXlq6nSU88\nCJwCDldVa9fS+DjwIeD5NWNtXk8BX09yNMneZqyN63kVsAJ8tkmHfSrJhYx5LW0I6J1XvV/PrTpu\nlORlwBeBD1TVr9e+1qb1VNUfqupaejvb65O8Zt3rrVlLkncAp6rq6Eb3tGk9jTc1fz5vp5fee/Pa\nF1u0nvOB1wGfrKrrgN+wLr0yjrW0IaAP1Yy6hZ5OsgugeTw14/kMLckF9IL5war6UjPc2vUAVNWv\ngG/Se6+jrWt5I/C3SR4HvgDckORfae96qKqnmsdTwJeB62nnep4Enmz+BQhwH70AP9a1tCGgT74Z\n9WwcAvY0z/fQy0XPvSQBPg08WlUfW/NS69aTZEeSi5vnL6X3XsBjtHAtAFV1V1VdVlW76f09+UZV\nvZeWrifJhUkuWn0OvA14iBaup6pOAr9IclUzdCPwCONey6zfLBjyDYVbgf8C/hv4yKzns4X5fx44\nAfye3m/qO4A/p/fm1THg68D2Wc9zyLW8id4/C38MPNh83drG9QCvBX7YrOUh4J+a8datpc/a3sKf\n3hRt5XqAvwR+1Hw9vPp3v8XruRZYbv5/+3fgknGvxUpRSeqINqRcJElDMKBLUkcY0CWpIwzoktQR\nBnRJ6ggDuiR1hAFdkjrCgC5JHfH/kzC8fGZ6v2sAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7f45ee6a2c50>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[ 1.  0.  0. ...,  0.  0.  0.]\n",
      " [ 0.  1.  0. ...,  0.  0.  0.]\n",
      " [ 0.  0.  1. ...,  0.  0.  0.]\n",
      " ..., \n",
      " [ 0.  0.  0. ...,  1.  0.  0.]\n",
      " [ 0.  0.  0. ...,  0.  1.  0.]\n",
      " [ 0.  0.  0. ...,  0.  0.  1.]]\n"
     ]
    }
   ],
   "source": [
    "def dotplot(dna_x, dna_y):\n",
    "    dotplot_matrix = np.zeros((len(dna_y), len(dna_x)), np.int)\n",
    "    for x_index, x_value in enumerate(dna_x):\n",
    "        for y_index, y_value in enumerate(dna_y):\n",
    "            if x_value == y_value:\n",
    "                dotplot_matrix[y_index,x_index] = 1\n",
    "    return dotplot_matrix\n",
    "\n",
    "dna_x = 'ATTGCAGCTTAAGGAATCGTGCAGATTAAAGGCACCACGAATTAAGACCAGGGACATAA'\n",
    "dna_y = 'ATTGCAGCGGAAGGAATACTGCAGGTTAAAGTCACCAGGAATTCAGACCAGTTTCATAA'\n",
    "\n",
    "print(dotplot(dna_x, dna_y))\n",
    "\n",
    "\n",
    "#dna_y = 'ATTGCAGCCTAAGGAATCGTGCAGATTAAACTCACCACGAATTAAGACCAGGGACATAT'\n",
    "p = dotplot(dna_x, dna_y)\n",
    "from matplotlib.pyplot import plot, show\n",
    "p = np.eye(60) # test\n",
    "x = [i for i in range(p.shape[1]) for j in range(p.shape[0]) if p[i,j] == 1]\n",
    "y = [j for i in range(p.shape[1]) for j in range(p.shape[0]) if p[i,j] == 1]\n",
    "plot(x, y, 'ro')\n",
    "show()\n",
    "\n",
    "# Test\n",
    "M = dotplot_list_of_lists(dna_x, dna_y)\n",
    "#for i in range(p.shape[0]):\n",
    "    #for j in range(p.shape[1]):\n",
    "        #assert p[i,j] == int(M[i][j]), '%s vs %s' % (p[i,j], int(M[i][j]))\n",
    "print(p)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on built-in function print in module builtins:\n",
      "\n",
      "print(...)\n",
      "    print(value, ..., sep=' ', end='\\n', file=sys.stdout, flush=False)\n",
      "    \n",
      "    Prints the values to a stream, or to sys.stdout by default.\n",
      "    Optional keyword arguments:\n",
      "    file:  a file-like object (stream); defaults to the current sys.stdout.\n",
      "    sep:   string inserted between values, default a space.\n",
      "    end:   string appended after the last value, default a newline.\n",
      "    flush: whether to forcibly flush the stream.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(print)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on function eye in module numpy.lib.twodim_base:\n",
      "\n",
      "eye(N, M=None, k=0, dtype=<class 'float'>)\n",
      "    Return a 2-D array with ones on the diagonal and zeros elsewhere.\n",
      "    \n",
      "    Parameters\n",
      "    ----------\n",
      "    N : int\n",
      "      Number of rows in the output.\n",
      "    M : int, optional\n",
      "      Number of columns in the output. If None, defaults to `N`.\n",
      "    k : int, optional\n",
      "      Index of the diagonal: 0 (the default) refers to the main diagonal,\n",
      "      a positive value refers to an upper diagonal, and a negative value\n",
      "      to a lower diagonal.\n",
      "    dtype : data-type, optional\n",
      "      Data-type of the returned array.\n",
      "    \n",
      "    Returns\n",
      "    -------\n",
      "    I : ndarray of shape (N,M)\n",
      "      An array where all elements are equal to zero, except for the `k`-th\n",
      "      diagonal, whose values are equal to one.\n",
      "    \n",
      "    See Also\n",
      "    --------\n",
      "    identity : (almost) equivalent function\n",
      "    diag : diagonal 2-D array from a 1-D array specified by the user.\n",
      "    \n",
      "    Examples\n",
      "    --------\n",
      "    >>> np.eye(2, dtype=int)\n",
      "    array([[1, 0],\n",
      "           [0, 1]])\n",
      "    >>> np.eye(3, k=1)\n",
      "    array([[ 0.,  1.,  0.],\n",
      "           [ 0.,  0.,  1.],\n",
      "           [ 0.,  0.,  0.]])\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(np.eye)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "class DNA(str):\n",
    "    \"\"\"\n",
    "    Class for representing DNA sequences as strings\n",
    "    \"\"\"\n",
    "    basecomplement = {'A': 'T', 'C': 'G', 'T': 'A', 'G': 'C', 'U': 'A'}\n",
    "    dna_codon_table = {\"TTT\":\"F\", \"TTC\":\"F\", \"TTA\":\"L\", \"TTG\":\"L\",\n",
    "                       \"TCT\":\"S\", \"TCC\":\"S\", \"TCA\":\"S\", \"TCG\":\"S\",\n",
    "                       \"TAT\":\"Y\", \"TAC\":\"Y\", \"TAA\":\"STOP\", \"TAG\":\"STOP\",\n",
    "                       \"TGT\":\"C\", \"TGC\":\"C\", \"TGA\":\"STOP\", \"TGG\":\"W\",\n",
    "                       \"CTT\":\"L\", \"CTC\":\"L\", \"CTA\":\"L\", \"CTG\":\"L\",\n",
    "                       \"CCT\":\"P\", \"CCC\":\"P\", \"CCA\":\"P\", \"CCG\":\"P\",\n",
    "                       \"CAT\":\"H\", \"CAC\":\"H\", \"CAA\":\"Q\", \"CAG\":\"Q\",\n",
    "                       \"CGT\":\"R\", \"CGC\":\"R\", \"CGA\":\"R\", \"CGG\":\"R\",\n",
    "                       \"ATT\":\"I\", \"ATC\":\"I\", \"ATA\":\"I\", \"ATG\":\"M\",\n",
    "                       \"ACT\":\"T\", \"ACC\":\"T\", \"ACA\":\"T\", \"ACG\":\"T\",\n",
    "                       \"AAT\":\"N\", \"AAC\":\"N\", \"AAA\":\"K\", \"AAG\":\"K\",\n",
    "                       \"AGT\":\"S\", \"AGC\":\"S\", \"AGA\":\"R\", \"AGG\":\"R\",\n",
    "                       \"GTT\":\"V\", \"GTC\":\"V\", \"GTA\":\"V\", \"GTG\":\"V\",\n",
    "                       \"GCT\":\"A\", \"GCC\":\"A\", \"GCA\":\"A\", \"GCG\":\"A\",\n",
    "                       \"GAT\":\"D\", \"GAC\":\"D\", \"GAA\":\"E\", \"GAG\":\"E\",\n",
    "                       \"GGT\":\"G\", \"GGC\":\"G\", \"GGA\":\"G\", \"GGG\":\"G\"}\n",
    "    def __init__(self, seq):\n",
    "        \"\"\"Create DNA object to string seq.\"\"\"\n",
    "        if set(seq.upper()).issubset({'A','T','C','G','N'}):\n",
    "            self.seq = seq.upper()\n",
    "        else:\n",
    "            raise ValueError(\"DNA has illegal character\")\n",
    "    \n",
    "    def transcribe(self):\n",
    "        \"\"\"\n",
    "        Transcribes a DNA sequence by replacing thymine (T) with uracil (U)\n",
    "        \"\"\"\n",
    "        return self.seq.replace('T', 'U')\n",
    "    \n",
    "    def complement(self):\n",
    "        \"\"\"\n",
    "        Finds the complement of the sequence by list comprehension using dictionary of complement values\n",
    "        \"\"\"\n",
    "        basecomplement = {'A': 'T', 'C': 'G', 'T': 'A', 'G': 'C', 'U': 'A', 'N': 'N'}\n",
    "        bases = list(self)\n",
    "        comp = [basecomplement[base] for base in bases]\n",
    "        return ''.join(comp)\n",
    "\n",
    "    def reverse(self):\n",
    "        \"\"\"\n",
    "        Reverses a sequence, used to get the 5' orientation of the complement strand of a sequence\n",
    "        \"\"\"\n",
    "        return self[::-1]\n",
    "\n",
    "    def gc_content(self):\n",
    "        \"\"\"\n",
    "        Calculates the GC content of a sequence, which is important to know in molecular biology\n",
    "        \"\"\"\n",
    "        gc_percent = (self.count('G') + self.count('C')) / len(self)\n",
    "        return gc_percent * 100\n",
    "\n",
    "    def codons(self, rf=0):\n",
    "        \"\"\"\n",
    "        Finds the list of codons in a sequence in a given reading frame.\n",
    "\n",
    "        In a sequence there are 6 possible reading frames, 3 on the forward strand, 3 on the complement.\n",
    "        Arguments: sequence for splitting in to codons\n",
    "        rf: reading frame. 0 = 1st forward frame; 1 = 2nd forward frame; 2 = 3rd forward frame\n",
    "            3 = 1st complement frame; 4 = 2nd complement frame; 5 = 3rd complement frame\n",
    "        Returns: list of codons in sequence for specified reading frame\n",
    "        \"\"\"\n",
    "        if rf == 0:\n",
    "            stoppoint = len(self) - (len(self) % 3)\n",
    "            return [self[i:i+3] for i in range(rf, stoppoint, 3)]\n",
    "        if rf == 1:\n",
    "            stoppoint = len(sequence[rf:]) - (len(self[rf:]) % 3)\n",
    "            return [self[i:i+3] for i in range(rf, stoppoint, 3)]\n",
    "        if rf == 2:\n",
    "            stoppoint = len(self[2:]) - (len(self[2:]) % 3)\n",
    "            return [self[i:i+3] for i in range(rf, stoppoint, 3)]\n",
    "        if rf == 3 or rf == 4 or rf == 5:\n",
    "            new_s = complement(self)\n",
    "            stoppoint = len(new_s[rf-3:]) - (len(new_s[rf-3:]) % 3)\n",
    "            return [new_s[i:i+3] for i in range(rf-3, stoppoint, 3)]\n",
    "        else:\n",
    "            raise ValueError(\"Reading frame rf must be a value from 0-5\")\n",
    "    \n",
    "    def translate(self):\n",
    "        \"\"\"\n",
    "        Translates an mRNA sequence into a protein sequence using the mRNA codon look up dictionary\n",
    "        \"\"\"\n",
    "        protein = \"\"\n",
    "        for i in range(0, len(self) - (len(self) % 3), 3):\n",
    "            symbol = dna_codon_table[self[i:i+3]]\n",
    "            if symbol == \"STOP\":\n",
    "                break\n",
    "            else:\n",
    "                protein += symbol\n",
    "        return protein\n",
    "    \n",
    "    def six_frames(self):\n",
    "        \"\"\"\n",
    "        Translate sequence into protein in all 6 reading frames,\n",
    "        prints the mrna with complement and the translated 6 frames aligned\n",
    "        with the nucleotides\n",
    "        \"\"\"\n",
    "        for i in reversed(range(3)):\n",
    "            protein = \" \"*i\n",
    "            for j in self.codons(self, rf=i):\n",
    "                if not dna_codon_table[j] == \"STOP\":\n",
    "                    protein += ((\" \") + dna_codon_table[j] + \" \")\n",
    "                else:\n",
    "                    protein += ' * '\n",
    "            print(protein)\n",
    "        print(self)\n",
    "        print(complement(self))\n",
    "        for i in range(3, 6):\n",
    "            protein = \" \"*(i-3)\n",
    "            for j in self.codons(self, rf=i):\n",
    "                if not dna_codon_table[j] == \"STOP\":\n",
    "                    protein += ((\" \") + dna_codon_table[j] + \" \")\n",
    "                else:\n",
    "                    protein += ' * '\n",
    "            print(protein)\n",
    "    \n",
    "    def __str__(self):\n",
    "        return \"DNA = {} \\n RNA = {} \\n Protein = {}\".format(self.seq, self.transcribe(), self.translate())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "dna_codon_table = {\"TTT\":\"F\", \"TTC\":\"F\", \"TTA\":\"L\", \"TTG\":\"L\",\n",
    "                   \"TCT\":\"S\", \"TCC\":\"S\", \"TCA\":\"S\", \"TCG\":\"S\",\n",
    "                   \"TAT\":\"Y\", \"TAC\":\"Y\", \"TAA\":\"STOP\", \"TAG\":\"STOP\",\n",
    "                   \"TGT\":\"C\", \"TGC\":\"C\", \"TGA\":\"STOP\", \"TGG\":\"W\",\n",
    "                   \"CTT\":\"L\", \"CTC\":\"L\", \"CTA\":\"L\", \"CTG\":\"L\",\n",
    "                   \"CCT\":\"P\", \"CCC\":\"P\", \"CCA\":\"P\", \"CCG\":\"P\",\n",
    "                   \"CAT\":\"H\", \"CAC\":\"H\", \"CAA\":\"Q\", \"CAG\":\"Q\",\n",
    "                   \"CGT\":\"R\", \"CGC\":\"R\", \"CGA\":\"R\", \"CGG\":\"R\",\n",
    "                   \"ATT\":\"I\", \"ATC\":\"I\", \"ATA\":\"I\", \"ATG\":\"M\",\n",
    "                   \"ACT\":\"T\", \"ACC\":\"T\", \"ACA\":\"T\", \"ACG\":\"T\",\n",
    "                   \"AAT\":\"N\", \"AAC\":\"N\", \"AAA\":\"K\", \"AAG\":\"K\",\n",
    "                   \"AGT\":\"S\", \"AGC\":\"S\", \"AGA\":\"R\", \"AGG\":\"R\",\n",
    "                   \"GTT\":\"V\", \"GTC\":\"V\", \"GTA\":\"V\", \"GTG\":\"V\",\n",
    "                   \"GCT\":\"A\", \"GCC\":\"A\", \"GCA\":\"A\", \"GCG\":\"A\",\n",
    "                   \"GAT\":\"D\", \"GAC\":\"D\", \"GAA\":\"E\", \"GAG\":\"E\",\n",
    "                   \"GGT\":\"G\", \"GGC\":\"G\", \"GGA\":\"G\", \"GGG\":\"G\"}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "dna1 = rna1.replace(\"U\", \"T\")\n",
    "dna1 = DNA(dna1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATGTACACTATCACTCTCGCCCCCACCACGTGGGGCAGCTTACAAACCAATGGGTCAGAGTGGCACACCTCAGGCGAGCACGTCTCGATGAACAGGTGA\n"
     ]
    }
   ],
   "source": [
    "print(dna1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'AGTGGACAAGTAGCTCTGCACGAGCGGACTCCACACGGTGAGACTGGGTAACCAAACATTCGACGGGGTGCACCACCCCCGCTCTCACTATCACATGTA'"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna1.reverse()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "codons() got multiple values for argument 'rf'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-31-cbc9d5e26344>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mdna1\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msix_frames\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m<ipython-input-24-cf11b3cfbc6d>\u001b[0m in \u001b[0;36msix_frames\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    102\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mreversed\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m3\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    103\u001b[0m             \u001b[0mprotein\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\" \"\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 104\u001b[0;31m             \u001b[0;32mfor\u001b[0m \u001b[0mj\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcodons\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrf\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    105\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mdna_codon_table\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mj\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m\"STOP\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    106\u001b[0m                     \u001b[0mprotein\u001b[0m \u001b[0;34m+=\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\" \"\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mdna_codon_table\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mj\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m\" \"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: codons() got multiple values for argument 'rf'"
     ]
    }
   ],
   "source": [
    "dna1.six_frames()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " |AUGUACACUAUCACUCUCGCCCCCACCACGUGGGGCAGCUUACAAACCAAUGGGUCAGAGUGGCACACCUCAGGCGAGCACGUCUCGAUGAACAGGUGA\n",
      "-----------------------------------------------------------------------------------------------------\n",
      "A|•   • •  •  •           •  •        •    • •••  ••      • •     • •    •    •  •       •  •• •    •\n",
      "U| • •    • •   • •             •        ••         •   •     •        •            • •   •       •  \n",
      "G|  •               •          • ••••  •             •••   • • ••         •• • •   •    •  •    •• • \n",
      "U| • •    • •   • •             •        ••         •   •     •        •            • •   •       •  \n",
      "A|•   • •  •  •           •  •        •    • •••  ••      • •     • •    •    •  •       •  •• •    •\n",
      "C|     • •   • • • • ••••• •• •      •  •   •   ••       •       • • •• •   •   • •  • •      •      \n",
      "A|•   • •  •  •           •  •        •    • •••  ••      • •     • •    •    •  •       •  •• •    •\n",
      "C|     • •   • • • • ••••• •• •      •  •   •   ••       •       • • •• •   •   • •  • •      •      \n",
      "U| • •    • •   • •             •        ••         •   •     •        •            • •   •       •  \n",
      "A|•   • •  •  •           •  •        •    • •••  ••      • •     • •    •    •  •       •  •• •    •\n",
      "U| • •    • •   • •             •        ••         •   •     •        •            • •   •       •  \n",
      "C|     • •   • • • • ••••• •• •      •  •   •   ••       •       • • •• •   •   • •  • •      •      \n",
      "A|•   • •  •  •           •  •        •    • •••  ••      • •     • •    •    •  •       •  •• •    •\n",
      "C|     • •   • • • • ••••• •• •      •  •   •   ••       •       • • •• •   •   • •  • •      •      \n",
      "U| • •    • •   • •             •        ••         •   •     •        •            • •   •       •  \n",
      "C|     • •   • • • • ••••• •• •      •  •   •   ••       •       • • •• •   •   • •  • •      •      \n",
      "U| • •    • •   • •             •        ••         •   •     •        •            • •   •       •  \n",
      "C|     • •   • • • • ••••• •• •      •  •   •   ••       •       • • •• •   •   • •  • •      •      \n",
      "G|  •               •          • ••••  •             •••   • • ••         •• • •   •    •  •    •• • \n",
      "C|     • •   • • • • ••••• •• •      •  •   •   ••       •       • • •• •   •   • •  • •      •      \n",
      "C|     • •   • • • • ••••• •• •      •  •   •   ••       •       • • •• •   •   • •  • •      •      \n",
      "C|     • •   • • • • ••••• •• •      •  •   •   ••       •       • • •• •   •   • •  • •      •      \n",
      "C|     • •   • • • • ••••• •• •      •  •   •   ••       •       • • •• •   •   • •  • •      •      \n",
      "C|     • •   • • • • ••••• •• •      •  •   •   ••       •       • • •• •   •   • •  • •      •      \n",
      "A|•   • •  •  •           •  •        •    • •••  ••      • •     • •    •    •  •       •  •• •    •\n",
      "C|     • •   • • • • ••••• •• •      •  •   •   ••       •       • • •• •   •   • •  • •      •      \n",
      "C|     • •   • • • • ••••• •• •      •  •   •   ••       •       • • •• •   •   • •  • •      •      \n",
      "A|•   • •  •  •           •  •        •    • •••  ••      • •     • •    •    •  •       •  •• •    •\n",
      "C|     • •   • • • • ••••• •• •      •  •   •   ••       •       • • •• •   •   • •  • •      •      \n",
      "G|  •               •          • ••••  •             •••   • • ••         •• • •   •    •  •    •• • \n",
      "U| • •    • •   • •             •        ••         •   •     •        •            • •   •       •  \n",
      "G|  •               •          • ••••  •             •••   • • ••         •• • •   •    •  •    •• • \n",
      "G|  •               •          • ••••  •             •••   • • ••         •• • •   •    •  •    •• • \n",
      "G|  •               •          • ••••  •             •••   • • ••         •• • •   •    •  •    •• • \n",
      "G|  •               •          • ••••  •             •••   • • ••         •• • •   •    •  •    •• • \n",
      "C|     • •   • • • • ••••• •• •      •  •   •   ••       •       • • •• •   •   • •  • •      •      \n",
      "A|•   • •  •  •           •  •        •    • •••  ••      • •     • •    •    •  •       •  •• •    •\n",
      "G|  •               •          • ••••  •             •••   • • ••         •• • •   •    •  •    •• • \n",
      "C|     • •   • • • • ••••• •• •      •  •   •   ••       •       • • •• •   •   • •  • •      •      \n",
      "U| • •    • •   • •             •        ••         •   •     •        •            • •   •       •  \n",
      "U| • •    • •   • •             •        ••         •   •     •        •            • •   •       •  \n",
      "A|•   • •  •  •           •  •        •    • •••  ••      • •     • •    •    •  •       •  •• •    •\n",
      "C|     • •   • • • • ••••• •• •      •  •   •   ••       •       • • •• •   •   • •  • •      •      \n",
      "A|•   • •  •  •           •  •        •    • •••  ••      • •     • •    •    •  •       •  •• •    •\n",
      "A|•   • •  •  •           •  •        •    • •••  ••      • •     • •    •    •  •       •  •• •    •\n",
      "A|•   • •  •  •           •  •        •    • •••  ••      • •     • •    •    •  •       •  •• •    •\n",
      "C|     • •   • • • • ••••• •• •      •  •   •   ••       •       • • •• •   •   • •  • •      •      \n",
      "C|     • •   • • • • ••••• •• •      •  •   •   ••       •       • • •• •   •   • •  • •      •      \n",
      "A|•   • •  •  •           •  •        •    • •••  ••      • •     • •    •    •  •       •  •• •    •\n",
      "A|•   • •  •  •           •  •        •    • •••  ••      • •     • •    •    •  •       •  •• •    •\n",
      "U| • •    • •   • •             •        ••         •   •     •        •            • •   •       •  \n",
      "G|  •               •          • ••••  •             •••   • • ••         •• • •   •    •  •    •• • \n",
      "G|  •               •          • ••••  •             •••   • • ••         •• • •   •    •  •    •• • \n",
      "G|  •               •          • ••••  •             •••   • • ••         •• • •   •    •  •    •• • \n",
      "U| • •    • •   • •             •        ••         •   •     •        •            • •   •       •  \n",
      "C|     • •   • • • • ••••• •• •      •  •   •   ••       •       • • •• •   •   • •  • •      •      \n",
      "A|•   • •  •  •           •  •        •    • •••  ••      • •     • •    •    •  •       •  •• •    •\n",
      "G|  •               •          • ••••  •             •••   • • ••         •• • •   •    •  •    •• • \n",
      "A|•   • •  •  •           •  •        •    • •••  ••      • •     • •    •    •  •       •  •• •    •\n",
      "G|  •               •          • ••••  •             •••   • • ••         •• • •   •    •  •    •• • \n",
      "U| • •    • •   • •             •        ••         •   •     •        •            • •   •       •  \n",
      "G|  •               •          • ••••  •             •••   • • ••         •• • •   •    •  •    •• • \n",
      "G|  •               •          • ••••  •             •••   • • ••         •• • •   •    •  •    •• • \n",
      "C|     • •   • • • • ••••• •• •      •  •   •   ••       •       • • •• •   •   • •  • •      •      \n",
      "A|•   • •  •  •           •  •        •    • •••  ••      • •     • •    •    •  •       •  •• •    •\n",
      "C|     • •   • • • • ••••• •• •      •  •   •   ••       •       • • •• •   •   • •  • •      •      \n",
      "A|•   • •  •  •           •  •        •    • •••  ••      • •     • •    •    •  •       •  •• •    •\n",
      "C|     • •   • • • • ••••• •• •      •  •   •   ••       •       • • •• •   •   • •  • •      •      \n",
      "C|     • •   • • • • ••••• •• •      •  •   •   ••       •       • • •• •   •   • •  • •      •      \n",
      "U| • •    • •   • •             •        ••         •   •     •        •            • •   •       •  \n",
      "C|     • •   • • • • ••••• •• •      •  •   •   ••       •       • • •• •   •   • •  • •      •      \n",
      "A|•   • •  •  •           •  •        •    • •••  ••      • •     • •    •    •  •       •  •• •    •\n",
      "G|  •               •          • ••••  •             •••   • • ••         •• • •   •    •  •    •• • \n",
      "G|  •               •          • ••••  •             •••   • • ••         •• • •   •    •  •    •• • \n",
      "C|     • •   • • • • ••••• •• •      •  •   •   ••       •       • • •• •   •   • •  • •      •      \n",
      "G|  •               •          • ••••  •             •••   • • ••         •• • •   •    •  •    •• • \n",
      "A|•   • •  •  •           •  •        •    • •••  ••      • •     • •    •    •  •       •  •• •    •\n",
      "G|  •               •          • ••••  •             •••   • • ••         •• • •   •    •  •    •• • \n",
      "C|     • •   • • • • ••••• •• •      •  •   •   ••       •       • • •• •   •   • •  • •      •      \n",
      "A|•   • •  •  •           •  •        •    • •••  ••      • •     • •    •    •  •       •  •• •    •\n",
      "C|     • •   • • • • ••••• •• •      •  •   •   ••       •       • • •• •   •   • •  • •      •      \n",
      "G|  •               •          • ••••  •             •••   • • ••         •• • •   •    •  •    •• • \n",
      "U| • •    • •   • •             •        ••         •   •     •        •            • •   •       •  \n",
      "C|     • •   • • • • ••••• •• •      •  •   •   ••       •       • • •• •   •   • •  • •      •      \n",
      "U| • •    • •   • •             •        ••         •   •     •        •            • •   •       •  \n",
      "C|     • •   • • • • ••••• •• •      •  •   •   ••       •       • • •• •   •   • •  • •      •      \n",
      "G|  •               •          • ••••  •             •••   • • ••         •• • •   •    •  •    •• • \n",
      "A|•   • •  •  •           •  •        •    • •••  ••      • •     • •    •    •  •       •  •• •    •\n",
      "U| • •    • •   • •             •        ••         •   •     •        •            • •   •       •  \n",
      "G|  •               •          • ••••  •             •••   • • ••         •• • •   •    •  •    •• • \n",
      "A|•   • •  •  •           •  •        •    • •••  ••      • •     • •    •    •  •       •  •• •    •\n",
      "A|•   • •  •  •           •  •        •    • •••  ••      • •     • •    •    •  •       •  •• •    •\n",
      "C|     • •   • • • • ••••• •• •      •  •   •   ••       •       • • •• •   •   • •  • •      •      \n",
      "A|•   • •  •  •           •  •        •    • •••  ••      • •     • •    •    •  •       •  •• •    •\n",
      "G|  •               •          • ••••  •             •••   • • ••         •• • •   •    •  •    •• • \n",
      "G|  •               •          • ••••  •             •••   • • ••         •• • •   •    •  •    •• • \n",
      "U| • •    • •   • •             •        ••         •   •     •        •            • •   •       •  \n",
      "G|  •               •          • ••••  •             •••   • • ••         •• • •   •    •  •    •• • \n",
      "A|•   • •  •  •           •  •        •    • •••  ••      • •     • •    •    •  •       •  •• •    •\n"
     ]
    }
   ],
   "source": [
    "def delta(x,y):\n",
    "    return 0 if x == y else 1\n",
    "\n",
    "def M(seq1,seq2,i,j,k):\n",
    "    return sum(delta(x,y) for x,y in zip(seq1[i:i+k],seq2[j:j+k]))\n",
    "\n",
    "def makeMatrix(seq1,seq2,k):\n",
    "    n = len(seq1)\n",
    "    m = len(seq2)\n",
    "    return [[M(seq1,seq2,i,j,k) for j in range(m-k+1)] for i in range(n-k+1)]\n",
    "\n",
    "def plotMatrix(M,t, seq1, seq2, nonblank = u'\\u2022', blank = ' '):\n",
    "    print(' |' + seq2)\n",
    "    print('-'*(2 + len(seq2)))\n",
    "    for label,row in zip(seq1,M):\n",
    "        line = ''.join(nonblank if s < t else blank for s in row)\n",
    "        print(label + '|' + line)\n",
    "        \n",
    "def dotplot(seq1,seq2,k = 1,t = 1):\n",
    "    M = makeMatrix(seq1,seq2,k)\n",
    "    plotMatrix(M, t, seq1,seq2)\n",
    "    \n",
    "seqx = \"ACCTGAGCTCACCTGAGTTA\"\n",
    "seqy = \"ACCTGAGCTCACCTGAGTTA\"\n",
    "import matplotlib.pyplot as plt\n",
    "dotplot(rna1,rna1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from ipywidgets import interact\n",
    "from IPython.display import HTML, display"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from ipywidgets import interact\n",
    "from IPython.display import HTML, display\n",
    "@interact(sequence=sequences,kmer=\"\" \n",
    ")\n",
    "def kmer_present(sequence=\"\", kmer = \"CCGATTCG\"):\n",
    "    \"\"\"This is an example of defining a function with keyword arguments:\n",
    "    \n",
    "    sequence and kmer are variables in the function kmer_present. \n",
    "    \n",
    "    These variables have DEFAULT VALUES (\"\" and \"CCGATTCG\" respectively) that will be used if no values are provided\n",
    "    explicitly.\n",
    "    \"\"\"\n",
    "    \n",
    "    # Here we are going to create a string with HTML markup. We will use the string format() method to insert\n",
    "    # our values into the string.\n",
    "    \n",
    "    # create a variable to store the length of the sequence\n",
    "\n",
    "    seq_length = len(sequence) # replace this with actual value\n",
    "    \n",
    "    # create a variable to store the answer of whether the kmer is in the sequence\n",
    "    s = '<h3>KMER Values:</h3><table>\\n'\n",
    "    s += '<tr><td>{0}</td><td>{1}</td></tr>\\n'.format(\"sequence length\",seq_length)\n",
    "    try:\n",
    "        kmercount=kmer_count(sequence, kmer)\n",
    "        if kmercount == 0:\n",
    "            high_sequence = sequence\n",
    "        else:\n",
    "            high_sequence = sequence.replace(kmer,'<font color=\"green\">%s</font>'%kmer)\n",
    "        \n",
    "        s+= '<tr><td>{0}</td><td>{1}</td></tr>\\n'.format(\"kmer: \"+kmer.upper(),kmercount)\n",
    "    except ValueError:\n",
    "        s+= '<tr><td>{0}</td><td>{1}</td></tr>\\n'.format(\"kmer: \"+kmer.upper(),\"Not a valid kmer\")\n",
    "        high_sequence = sequence\n",
    "    s += '<tr><td>{0}</td><td>{1}</td></tr>'.format(\"sequence \", high_sequence)    \n",
    "    s += '</table>'\n",
    "\n",
    "    display(HTML(s))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
